python/ray/workflow/workflow_executor.py

```diff
@@ -0,0 +1,166 @@
+
+
+logger = logging.getLogger(__name__)
+
+
+async def execute_workflow(
+    job_id: str,
+    context: WorkflowStepContext,
+    state: WorkflowExecutionState,
+    wf_store: workflow_storage.WorkflowStorage,
+) -> None:
+    """The core logic of execution a workflow.
+
+    This function is responsible for:
+
+    - Dependency resolving.
+    - Task scheduling.
+    - Reference counting.
+    - Garbage collection.
+    - Continuation handling and scheduling.
+    - Error handling.
+    - Responding callbacks.
+
+    Args:
+        job_id: The Ray JobID for logging properly.
+        context: The context of workflow execution.
+        wf_store: The store for the workflow.
+
+    # TODO(suquark): move job_id inside context
+    """
+    workflow_id = context.workflow_id
+
+    # ------------ init ------------
+    state.construct_runtime_deps(state.output_task_id)
+    state.init_context(context)
+
+    completion_queue = asyncio.Queue()
+
+    while state.is_running():
+        # ------------ submit task ------------
+        while True:
+            if state.stopping:
+                raise WorkflowCancellationError(workflow_id)
+            task_id = state.pop_frontier_to_run()
+            if not task_id:
+                break
+
+            baked_inputs = _BakedWorkflowInputs(
+                args=state.task_input_args[task_id],
+                workflow_refs=[
+                    state.get_input(d) for d in state.in_dependency[task_id]
+                ],
+            )
+            task_options = state.task_options[task_id]
+            executor = get_step_executor(task_options)
+            metadata_ref, output_ref = executor(
+                state.task_func_body[task_id],
+                state.task_context[task_id],
+                job_id,
+                task_id,
+                baked_inputs,
+                task_options,
+            )
+            # The input workflow is not a reference to an executed workflow.
+            future = asyncio.wrap_future(metadata_ref.future())
+            future.add_done_callback(completion_queue.put_nowait)
+
+            state.insert_running_frontier(future, WorkflowRef(task_id, ref=output_ref))
+            state.task_execution_metadata[task_id] = TaskExecutionMetadata(
+                submit_time=time.time()
+            )
+            if task_id in state.continuation_root:
+                if state.task_options[task_id].checkpoint:
+                    wf_store.update_continuation_output_link(
+                        state.continuation_root[task_id], task_id
+                    )
+            else:
+                # update reference counting if the task is not a continuation
+                for c in state.in_dependency[task_id]:
+                    state.reference_set[c].remove(task_id)
+                    if not state.reference_set[c]:
+                        del state.reference_set[c]
+                        state.free_outputs.add(c)
```

Comment:
Might be good to wrap this in a method on state.

Reply:
now it is wrapped in a method in `workflow_executor`